src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs

```diff
@@ -729,5 +729,153 @@ class Test
                 workspace.Options = originalOptions;
             }
         }
+
+        [WorkItem(23590, "https://github.com/dotnet/roslyn/issues/23590")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void TypeImplementsIEnumerableOfType()
+        {
+            var markup = @"
+using System.Collections.Generic;
+
+public class Class1
+{
+  public void Method()
+  {
+    Container $$
+  }
+}
+
+public class Container : ContainerBase { }
+public class ContainerBase : IEnumerable<ContainerBase> { }
+";
+            await VerifyItemExistsAsync(markup, "container");
+        }
+
+        [WorkItem(23590, "https://github.com/dotnet/roslyn/issues/23590")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void TypeImplementsIEnumerableOfType2()
+        {
+            var markup = @"
+using System.Collections.Generic;
+
+public class Class1
+{
+  public void Method()
+  {
+     Container $$
+  }
+}
+
+public class ContainerBase : IEnumerable<Container> { }
+public class Container : ContainerBase { }
+";
+            await VerifyItemExistsAsync(markup, "container");
+        }
+
+        [WorkItem(23590, "https://github.com/dotnet/roslyn/issues/23590")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void TypeImplementsIEnumerableOfType3()
+        {
+            var markup = @"
+using System.Collections.Generic;
+
+public class Class1
+{
+  public void Method()
+  {
+     Container $$
+  }
+}
+
+public class Container : IEnumerable<Container> { }
+";
+            await VerifyItemExistsAsync(markup, "container");
+        }
+
+        [WorkItem(23590, "https://github.com/dotnet/roslyn/issues/23590")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void TypeImplementsIEnumerableOfType4()
+        {
+            var markup = @"
+using System.Collections.Generic;
+
+public class Class1
+{
+  public void Method()
+  {
+     TaskType $$
+  }
+}
+
+public class ContainerBase : IEnumerable<Container> { }
+public class Container : ContainerBase { }
+public class TaskType : Task<Container> { }
```

Comment:
❓ Did you mean to have `using System.Threading.Tasks;`? Same applies to the other tests that reference `Task<T>`.

Reply:
Fixed.